### PR TITLE
Feature/formbuilder

### PR DESCRIPTION
**Pull request by trasher**
See https://glpi-developer-documentation.readthedocs.io/en/feature-formbuilder/devapi/forms.html

### CODE REVIEW

**Comment by trasher** [comment]
Also rely on https://github.com/glpi-project/glpi/pull/5569 and https://github.com/glpi-project/glpi/pull/5557 features.

File: source/devapi/forms.rst [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+* ``getEditForm()``, ``getAddForm()`` (both aliases to ``getForm()``) will give you a form definition for the current object configuration,
+* ``getFormFields()`` describes each form elements (and their order in the form). Some :ref:`defaults are provided <defaultfields>`, this is is good idea to cal ``parent::getFormFields()`` when overrided,
+* ``cleanFormFields()`` (optional) clean form fields that are not present in the database,
+* ``getFormHiddenFields()`` list of fields that would be hidden,
+* ``getFormFieldsToDrop()`` list of fields that should not be present at all on the form (but has been retrieved from database),
+* ``buildFormElements()`` is used on each element to get its configuration
+
+.. note::
+
+   Almost all of the abos methods do provide usefull default values. Do not forget to call the parent method if you want to override!
```
Comment (orthagh) [inline source/devapi/forms.rst]:
```suggestion
   Almost all of the above methods do provide useful default values. Do not forget to call the parent method if you want to override!
```

File: source/devapi/forms.rst [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+#. fields are loaded from the table columns,
+#. load fields configuration,
+#. elements to drop are handled,
+#. for each database column, a form element is built,
+#. if any mapping configuration is set in :ref:`fields mapping <mapped_fields>`, it will be handled.
+
+Based on field names, the form builder will automatically guess some configuration for you:
+
+* any field name containing 'date' will be configured with a date picker,
+* any foreign key field (name ends with ``_id`` or contains ``_id_``) will be configured ass an ajax dropdown, label will be retrieved from the Object, `add` and `list` buttons will be added. For locations, the `map` button will be added as well,
```
Comment (orthagh) [inline source/devapi/forms.rst]:
```suggestion
* any foreign key field (name ends with ``_id`` or contains ``_id_``) will be configured as an ajax dropdown, label will be retrieved from the Object, `add` and `list` buttons will be added. For locations, the `map` button will be added as well,
```

File: source/devapi/forms.rst [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+* fields names containing ``_cache`` will be set as hidden,
+* fields names ``content`` and ``comment`` will be configured as textareas,
+* fields names that begins with ``is_`` or contains ``_is_`` will set to ``yesno``.
```
Comment (orthagh) [inline source/devapi/forms.rst]:
```suggestion
* fields names that begins with ``is_`` or contains ``_is_`` will be set to ``yesno``.
```

File: source/devapi/forms.rst [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+      'serial' => [
+         'label'           => 'Serial',
+         'autofill'        => true,
+         'type'            => 'text',
+         'name'            => 'serial',
+         'autofocus'       => false
+      ]
+   ];
+
+The key is the element unique identitifer. It must be unique in each form.
```
Comment (orthagh) [inline source/devapi/forms.rst]:
```suggestion
The key is the element unique identifier. It must be unique in each form.
```

File: source/devapi/forms.rst [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+For purge:
+
+* ``with_check`` set to false to not require user confirmation, checkbox will not be displayed,
+* ``with_restore`` set to false to not display restore button.
+
+
+Fieldsets
+^^^^^^^^^
+
+By default, all form elements will take place in a songle fieldset. I fyou want to use several fieldsets inside your form, you will have to use the ``fieldset`` entry from the field configuration to tag your elements. Then you have to use the ``getFieldsets()`` method to declare your fieldsets:
```
Comment (orthagh) [inline source/devapi/forms.rst]:
```suggestion
By default, all form elements will take place in a single fieldset. If you want to use several fieldsets inside your form, you will have to use the ``fieldset`` entry from the field configuration to tag your elements. Then you have to use the ``getFieldsets()`` method to declare your fieldsets:
```

File: source/devapi/forms.rst [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+* ``title`` that will be displayed as ``fieldset/legend``,
+* ``subtitle`` (optional) will be shown along with the legend to provide extra informations,
+* ``elements`` will be automatically populated in the ``getForm`` method with fields tagged with current fieldset.
+
+If a fieldset is tagged in any element but does not exists, a warning will be displayed in logs.
+
+Fields mapping
+^^^^^^^^^^^^^^
+
+By default, any form element represents a database column, and will be handled separately from all others. In some cases, say when we want to display an itemtype link, we'll need several f those informations (``itemtype`` and ``items_id`` in our example). This example is based on ``ItemDisk`` configuration and use ``templates/elements/link.twig`` template.
```
Comment (orthagh) [inline source/devapi/forms.rst]: "we'll need several f those informations" <- you miss something in this phrase, but i can't figure it

**Pull request merged**